engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/FramePointerAnalysis.scala

```diff
@@ -0,0 +1,340 @@
+package org.enso.compiler.pass.analyse
+
+import org.enso.compiler.context.{
+  CompilerContext,
+  FramePointer,
+  InlineContext,
+  LocalScope,
+  ModuleContext
+}
+import org.enso.compiler.core.{CompilerError, IR}
+import org.enso.compiler.core.ir.expression.Application
+import org.enso.compiler.core.ir.{
+  CallArgument,
+  DefinitionArgument,
+  Expression,
+  Function,
+  Module,
+  Name,
+  ProcessingPass
+}
+import org.enso.compiler.core.Implicits.AsMetadata
+import org.enso.compiler.core.ir.MetadataStorage.MetadataPair
+import org.enso.compiler.core.ir.module.scope.Definition
+import org.enso.compiler.core.ir.module.scope.definition.Method
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.IRPass.IRMetadata
+import org.enso.compiler.pass.analyse.alias.{Graph, Info}
+
+/** This pass attaches [[FramePointer]] as metadata to all the IR elements that already
+  * have [[org.enso.compiler.pass.analyse.alias.Info.Occurrence]] attached.
+  */
+case object FramePointerAnalysis extends IRPass {
+
+  override type Metadata = FramePointerMeta
```

Comment:
`FramePointerMeta` seems to add one additional and unnecessary level of indirection. It'd be simpler if the object in the `Metadata` was directly `FramePointer`.

Refactoring of [various IRMetadata concepts](https://github.com/enso-org/enso/pull/10729/files#diff-bd9022d776b808f603e287826601f2d1dc9bf703a17972645ebb1ff7ab7de7ebR330) would however be very likely needed...